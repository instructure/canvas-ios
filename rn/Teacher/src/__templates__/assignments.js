//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* @flow */

import template, { type Template } from '../utils/template'

export const assignment: Template<Assignment> = template({
  id: '123456789',
  course_id: '987654321',
  name: 'learn to write code',
  description: 'a course that can take any beginner to professional in a matter of microseconds.  Harness the power of your favorite languages like PHP and Cold Fusion.',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
  created_at: '2017-03-17T19:15:25Z',
  updated_at: '2017-03-17T19:15:25Z',
  points_possible: 20,
  grading_type: 'points',
  assignment_group_id: '1376',
  peer_reviews: false,
  automatic_peer_reviews: false,
  position: 1,
  submission_types:
  [ 'online_text_entry',
    'online_url',
    'media_recording',
    'online_upload' ],
  needs_grading_count: 0,
  html_url: 'http://hostname.tld/whatever',
  overrides: [],
  grade_group_students_individually: false,
  group_category_id: null,
  has_overrides: false,
  only_visible_to_overrides: false,
  published: true,
  unpublishable: false,
  rubric_assessment: {},
  rubric_settings: null,
  rubric: null,
  anonymize_students: false,
})

export const assignmentGroup: Template<AssignmentGroup> = template({
  id: '1',
  name: 'Learn React Native',
  position: 1,
  assignments: [assignment({ assignment_group_id: '1' })],
  group_weight: 1,
  integration_data: {},
  sis_source_id: '',
})

export const assignmentDueDate: Template<AssignmentDate> = template({
  title: 'due',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
})

export const assignmentOverride: Template<AssignmentOverride> = template({
  id: '1',
  assignment_id: '1',
  student_ids: [],
  group_id: '1',
  course_section_id: '1',
  title: 'override',
  all_day: false,
  all_date_date: '',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
})
