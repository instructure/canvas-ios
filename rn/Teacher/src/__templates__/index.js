//
// This file is part of Canvas.
// Copyright (C) 2018-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

export * from '../redux/__templates__/app-state'
export * from './account-notification'
export * from './addressBook'
export * from './api'
export * from './assignments'
export * from './attachment'
export * from './conversations'
export * from './course'
export * from './discussion'
export * from './enrollments'
export * from './error'
export * from './external-tool'
export * from './file'
export * from './grading-periods'
export * from './group'
export * from './helm'
export * from './launch-definitions'
export * from './mediaComment'
export * from './quiz'
export * from './quizSubmission'
export * from './role'
export * from './rubric'
export * from './section'
export * from './session'
export * from './submissions'
export * from './tab'
export * from './users'
export * from './help-links'
export * from './dashboard'
export * from '../modules/assignee-picker/__template__/Assignee'
