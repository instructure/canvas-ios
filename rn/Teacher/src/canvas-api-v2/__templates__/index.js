// @flow
export * from './submission'
export * from './user'
export * from './assignment'
export * from './course'
export * from './group'
export * from './section'
export * from './groupSet'

// QUERIES
export * from '../queries/__templates__/SubmissionList'
export * from '../queries/__templates__/ContextCard'
