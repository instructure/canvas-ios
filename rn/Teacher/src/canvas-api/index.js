//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

export { default } from './apis/index'
export * from './apis/assignmentGroups'
export * from './apis/assignments'
export * from './apis/conversations'
export * from './apis/courses'
export * from './apis/enrollments'
export * from './apis/external-tools'
export * from './apis/groups'
export * from './apis/login'
export * from './apis/quizzes'
export * from './apis/submissions'
export * from './apis/users'
export * from './apis/media-comments'
export * from './apis/feature-flags'
export * from './apis/file-uploads'
export * from './apis/accounts'
export * from './apis/dashboard'

export * from './session'

export { default as httpClient, isAbort, httpCache } from './httpClient'
