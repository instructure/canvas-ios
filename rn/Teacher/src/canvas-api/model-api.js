//
// This file is part of Canvas.
// Copyright (C) 2018-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
export * from './fetch-props-for'
export * from './model'
export { httpCache } from './httpClient'

import httpClient, { httpCache } from './httpClient'
import parseLink from './utils/parse-link-header'
import {
  CourseModel,
  PageModel,
  ToDoModel,
} from './model'

type ApiPolicy = 'cache-only' | 'cache-and-network' | 'network-only'
type ApiOptions = {|
  policy: ApiPolicy,
  onStart?: ApiPromise<any> => void,
  onComplete?: ApiPromise<any> => void,
  onAnyComplete?: ApiPromise<any> => void,
  onError?: Error => void,
|}
type Paginated<T> = {|
  list: T,
  next: ?string,
  getNextPage: ?(() => ApiPromise<Paginated<T>>),
|}

const emptyArray = []
const emptyPaginated: Paginated<any> = {
  list: emptyArray,
  next: null,
  getNextPage: null,
}

export class API {
  options: ApiOptions

  constructor (options: ApiOptions) {
    this.options = options
  }

  cleanup () {
    this.options.onStart =
    this.options.onComplete =
    this.options.onAnyComplete =
    this.options.onError = undefined
  }

  get (url: string, config: ApiConfig = {}) {
    const { policy } = this.options
    const cached = httpCache.get(url, config)
    if (
      policy === 'network-only' ||
      (policy === 'cache-and-network' && cached.expiresAt < Date.now())
    ) {
      this.request(httpClient.get(url, config))
    }
    return cached.value || null
  }

  paginate<T> (url: string, config: ApiConfig = {}): Paginated<T> {
    const transform = config.transform
    config.transform = (value: any, response: ApiResponse<any>) => {
      const list = transform ? transform(value, response) : value
      const { current, first, next } = parseLink(response.headers.link) || {}
      const getNextPage = next && (() => {
        this.request(httpClient.get(next, config))
      })
      if (getNextPage && config.params && config.params.per_page >= 100) {
        getNextPage() // automatically request all pages
      }
      if (current && current !== first) {
        const base = httpCache.get(url, config).value
        if (base) {
          base.list = base.list.concat(list)
          base.next = next
          base.getNextPage = getNextPage
        }
        return null
      }
      return {
        list,
        next,
        getNextPage,
      }
    }

    const cached = httpCache.get(url, config)
    const { policy } = this.options
    if (
      policy === 'network-only' ||
      (policy === 'cache-and-network' && cached.expiresAt < Date.now())
    ) {
      this.request(httpClient.get(url, config))
    }
    // getNextPage won't survive serialization
    const value = cached.value
    if (value && value.next && !value.getNextPage && policy !== 'cache-only') {
      value.getNextPage = () => this.request(httpClient.get(value.next, config))
    }
    return value || emptyPaginated
  }

  post (url: string, data: *, config: ApiConfig = {}) {
    return this.request(httpClient.post(url, data, config))
  }

  put (url: string, data: *, config: ApiConfig = {}) {
    return this.request(httpClient.put(url, data, config))
  }

  delete (url: string, config: ApiConfig = {}) {
    return this.request(httpClient.delete(url, config))
  }

  request (promise: ApiPromise<any>) {
    const { onStart, onComplete, onError } = this.options
    onStart && onStart(promise)
    promise
      .catch(error => { onError && onError(error) })
      .then(() => { onComplete && onComplete(promise) })
    return promise
  }

  getCourseColor (courseID: string): string {
    const json: ?Object = this.get('users/self/colors')
    return json && json.custom_colors[`course_${courseID}`] || '#aaa'
  }

  getCourse (courseID: string): ?CourseModel {
    return this.get(`courses/${courseID}`, {
      params: {
        include: [ 'permissions', 'term', 'favorites', 'course_image', 'sections' ],
      },
      transform: (course: Course) => new CourseModel(course),
    })
  }

  getPages (context: CanvasContext, contextID: string): Paginated<PageModel[]> {
    return this.paginate(`${context}/${contextID}/pages`, {
      params: {
        per_page: 100,
      },
      transform: (pages: Page[]) => pages.map(page => new PageModel(page)),
    })
  }

  createPage (context: CanvasContext, contextID: string, parameters: PageParameters): ApiPromise<PageModel> {
    return this.post(`${context}/${contextID}/pages`, { wiki_page: parameters }, {
      transform: (page: Page) => new PageModel(page),
    })
  }

  getPage (context: CanvasContext, contextID: string, url: string): ?PageModel {
    return this.get(`${context}/${contextID}/pages/${url}`, {
      transform: (page: Page) => new PageModel(page),
    })
  }

  getFrontPage (contextID: string): ?PageModel {
    return this.get(`courses/${contextID}/front_page`, {
      transform: (page: Page) => new PageModel(page),
    })
  }

  updatePage (context: CanvasContext, contextID: string, url: string, parameters: PageParameters): ApiPromise<PageModel> {
    return this.put(`${context}/${contextID}/pages/${url}`, { wiki_page: parameters }, {
      transform: (page: Page) => new PageModel(page),
    })
  }

  deletePage (context: CanvasContext, contextID: string, url: string): ApiPromise<null> {
    return this.delete(`${context}/${contextID}/pages/${url}`)
  }

  getToDos (): Paginated<ToDoModel[]> {
    return this.paginate('users/self/todo', {
      transform: (todos: ToDoItem[]) => todos.map(todo => new ToDoModel(todo)),
    })
  }
}
