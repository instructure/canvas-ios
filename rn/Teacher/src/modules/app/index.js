//
// Copyright (C) 2017-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

export type AppId = 'student' | 'teacher' | 'parent'
export type App = {
  appId: AppId,
  filterCourse: (course: Course) => boolean,
}

const teacher = {
  appId: 'teacher',
  filterCourse: (course: Course): boolean => {
    if (course.access_restricted_by_date) return false
    const enrollments = course.enrollments
    if (!enrollments) return false
    return enrollments.some((e) =>
      [
        'teacher',
        'teacherenrollment',
        'designer',
        'ta',
      ].includes(e.type.toLowerCase())
    )
  },
}

const student = {
  appId: 'student',
  filterCourse: (course: Course): boolean => !course.access_restricted_by_date,
}

const parent = {
  appId: 'parent',
  filterCourse: (course: Course): boolean => true,
}

let current: App = teacher

const app = {
  setCurrentApp: (appId: AppId): void => {
    switch (appId) {
      case 'student':
        current = student
        break
      case 'teacher':
        current = teacher
        break
      case 'parent':
        current = parent
        break
    }
  },
  current: (): App => current,
}

export function isTeacher (): boolean {
  return app.current().appId === 'teacher'
}
export function isStudent (): boolean {
  return app.current().appId === 'student'
}

export function isParent (): boolean {
  return app.current().appId === 'parent'
}

export default app
