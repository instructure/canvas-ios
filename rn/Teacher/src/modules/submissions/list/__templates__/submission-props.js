//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import template, { type Template } from '../../../../utils/template'
import type { SubmissionDataProps } from '../submission-prop-types'

const templates = {
  ...require('../../../../__templates__/submissions'),
}

const submissionWithHistory = templates.submissionHistory()

// $FlowFixMe
export const submissionProps: Template<SubmissionDataProps> = template({
  grade: 'B-',
  name: submissionWithHistory.user.name,
  status: 'submitted',
  userID: submissionWithHistory.user.id,
  avatarURL: submissionWithHistory.user.avatar_url,
  sectionID: '1',
})
