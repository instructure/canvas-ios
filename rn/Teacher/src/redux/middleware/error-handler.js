//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import { Alert } from 'react-native'
import type { MiddlewareAPI } from 'redux'
import i18n from 'format-message'
import loginVerify from '../../common/login-verify'

export function alertError (error: any, alertTitle?: string, callback?: Function): void {
  const title = alertTitle || defaultErrorTitle()
  const message = parseErrorMessage(error)
  let buttons = [{ text: i18n('Dismiss'), onPress: () => { if (callback) callback() } }]
  Alert.alert(title, message, buttons)

  if (error instanceof Error) {
    console.warn(`Error: ${error.message}.  Stack:\n${error.stack}`)
  } else {
    console.warn(`Error: ${error}`)
  }
}

export function defaultErrorTitle (): string {
  return i18n('Unexpected Error')
}

export function defaultErrorMessage (): string {
  return i18n('There was an unexpected error. Please try again.')
}

let showingGlobalErrorAlert = false
const showGlobalErrorAlertIfNecessary = async (error: any) => {
  if (showingGlobalErrorAlert) return
  showingGlobalErrorAlert = true
  if (error.response != null) {
    error = error.response
  }

  let alertTitle

  alertError(error, alertTitle, () => {
    resetGlobalErrorAlert()
  })
}

export function resetGlobalErrorAlert () {
  showingGlobalErrorAlert = false
}

const errorHandlerMiddleware: MiddlewareAPI = () => {
  return next => action => {
    if (action.error) {
      let error = action.payload.error
      if (error && error.response && error.response.status === 401) {
        loginVerify().then((invalid) => {
          if (!invalid) {
            if (action.payload.handlesError) return next(action)
            showGlobalErrorAlertIfNecessary(error)
          }
          next(action)
        })
        return action // Bail out of this because of the promise. Next will be called after the promise has finished
      } else {
        if (action.payload.handlesError) return next(action)
        showGlobalErrorAlertIfNecessary(error)
      }
    }
    return next(action)
  }
}

export default errorHandlerMiddleware

export function parseErrorMessage (error: any): string {
  if (error && error.response) error = error.response

  if (typeof error === 'string') {
    return error
  }

  if (error && error.data && error.data.message) {
    return error.data.message
  } else if (error && error.data && error.data.errors && error.data.errors.length > 0) {
    if (typeof error.data.errors === 'string') {
      return error.data.errors
    }
    return error.data.errors
      .map(error => error.message)
      .map(message => message.replace(/\.+$/, ''))
      .join('. ')
  } else if (error && error.data && error.data.errors instanceof Object && Object.keys(error.data.errors).length > 0) {
    let data = error.data.errors
    let result = Object.keys(data)
      .map((key, index) => {
        let err = data[key]
        if (Array.isArray(err) && err.length > 0) err = err[0]
        if (typeof err === 'string') return err
        return err && err.message || ''
      })
      .join('. ')

    if (result.trim()) {
      return result
    }
  }

  if (error instanceof Error) {
    return error.message
  }

  return defaultErrorMessage()
}
