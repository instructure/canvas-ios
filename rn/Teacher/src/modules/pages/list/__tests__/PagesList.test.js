//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import { shallow } from 'enzyme'
import React from 'react'
import { alertError } from '../../../../redux/middleware/error-handler'
import { httpCache } from '../../../../canvas-api/model-api'
import * as template from '../../../../__templates__'
import app from '../../../app'
import ConnectedPagesList, { PagesList } from '../PagesList'

jest.mock('../../../../redux/middleware/error-handler', () => {
  return { alertError: jest.fn() }
})
jest.useFakeTimers()

describe('PagesList', () => {
  let props
  beforeEach(() => {
    jest.clearAllMocks()
    httpCache.clear()
    props = {
      courseID: '1',
      navigator: template.navigator(),
      courseColor: 'green',
      course: template.courseModel({ id: '1' }),
      pages: [template.pageModel()],
      isLoading: false,
      loadError: null,
      refresh: jest.fn(),
    }
  })

  it('gets courseColor, course, and pages from the model api', () => {
    const courseColor = 'green'
    const course = [ template.courseModel() ]
    const pages = [ template.pageModel() ]
    httpCache.handle('GET', 'users/self/colors', { custom_colors: { course_1: courseColor } })
    httpCache.handle('GET', 'courses/1', course)
    httpCache.handle('GET', 'courses/1/pages', { list: pages })
    const tree = shallow(<ConnectedPagesList courseID='1' />)
    expect(tree.find(PagesList).props()).toMatchObject({
      courseColor,
      course,
      pages,
    })
  })

  it('renders empty', () => {
    props.pages = []
    const tree = shallow(<PagesList {...props} />)
    expect(tree).toMatchSnapshot()
  })

  it('renders loading', () => {
    props.course = null
    props.pages = []
    props.isLoading = true
    const tree = shallow(<PagesList {...props} />)
    expect(tree).toMatchSnapshot()
  })

  it('renders front page row', () => {
    props.pages = [template.pageModel({ isFrontPage: true, title: 'Page 1' })]
    const tree = shallow(<PagesList {...props} />)
    const row = tree.find('FlatList').dive().find('FeatureRow').first()
    expect(row.prop('title')).toBe('Front Page')
    expect(row.prop('subtitle')).toBe('Page 1')
  })

  it('sorts the pages list', () => {
    const tree = shallow(<PagesList {...props} />)
    const pages = [
      template.pageModel({ title: 'Page B' }),
      template.pageModel({ isFrontPage: true, title: 'Page 1' }),
      template.pageModel({ title: 'Page A' }),
    ]
    tree.setProps({ pages })
    const sorted = tree.find('FlatList').prop('data')
    expect(sorted).toEqual([
      template.pageModel({ isFrontPage: true, title: 'Page 1' }),
      template.pageModel({ title: 'Page A' }),
      template.pageModel({ title: 'Page B' }),
    ])
    tree.setProps({ pages }) // not changed
    expect(tree.find('FlatList').prop('data')).toBe(sorted)
  })

  it('shows access icon for teachers', () => {
    app.setCurrentApp('teacher')
    const tree = shallow(<PagesList {...props} />)
    const row = tree.find('FlatList').dive().find('Row').first()
    expect(row.prop('renderImage')()).toMatchSnapshot()
  })

  it('alerts on new loadError', () => {
    const tree = shallow(<PagesList {...props} />)
    tree.setProps({ loadError: null })
    expect(alertError).not.toHaveBeenCalled()
    const loadError = new Error()
    tree.setProps({ loadError })
    expect(alertError).toHaveBeenCalledWith(loadError)
  })

  it('navigates to page details', () => {
    props.navigator = template.navigator({ show: jest.fn() })
    props.pages = [template.pageModel({ url: 'abc' })]
    const tree = shallow(<PagesList {...props} />)
    tree.find('FlatList').dive().find('Row').first()
      .simulate('Press', 'abc')
    expect(props.navigator.show).toHaveBeenCalledWith(
      '/courses/1/pages/abc',
      { modal: false }
    )
  })

  it('navigates to new page', () => {
    props.navigator = template.navigator({ show: jest.fn() })
    const tree = shallow(<PagesList {...props} />)
    tree.find('Screen').prop('rightBarButtons')
      .find(({ testID }) => testID === 'pages.list.add.button')
      .action()
    expect(props.navigator.show).toHaveBeenCalledWith(
      '/courses/1/pages/new',
      { modal: true, modalPresentationStyle: 'formsheet' }
    )
  })

  it('shows add button if permitted', () => {
    app.setCurrentApp('teacher')
    const tree = shallow(<PagesList {...props} />)
    expect(
      tree.prop('rightBarButtons').find(({ testID }) => testID === 'pages.list.add.button')
    ).toBeTruthy()
  })

  it('hides add button if not permitted', () => {
    app.setCurrentApp('student')
    const tree = shallow(<PagesList {...props} />)
    expect(tree.prop('rightBarButtons')).toBeFalsy()
  })

  it('shows front page', () => {
    props.navigator = template.navigator({
      traitCollection: (callback) => {
        callback({
          screen: {
            horizontal: 'regular',
          },
          window: {
            horizontal: 'regular',
          },
        })
      },
    })
    props.pages = [template.pageModel({ url: 'front-page', isFrontPage: true })]
    shallow(<PagesList {...props} />)
    expect(props.navigator.show).toHaveBeenLastCalledWith(
      '/courses/1/pages/front-page',
      { modal: false },
    )
  })

  it('does not show front page on deep link modal', () => {
    props.navigator = template.navigator({
      traitCollection: (callback) => {
        callback({
          screen: {
            horizontal: 'regular',
          },
          window: {
            horizontal: 'regular',
          },
        })
      },
      isModal: true,
    })
    props.pages = [template.pageModel({ url: 'front-page', isFrontPage: true })]
    shallow(<PagesList {...props} />)
    expect(props.navigator.show).not.toHaveBeenCalled()
  })

  it('shows placeholder if no front page', () => {
    props.navigator = template.navigator({
      traitCollection: (callback) => {
        callback({
          screen: {
            horizontal: 'regular',
          },
          window: {
            horizontal: 'regular',
          },
        })
      },
    })
    props.pages = [template.pageModel({ isFrontPage: false })]
    const course = props.course = template.courseModel({ id: '1' })
    shallow(<PagesList {...props} />)
    expect(props.navigator.show).toHaveBeenLastCalledWith(
      '/courses/1/placeholder',
      {},
      { courseColor: 'green', course: course.raw },
    )
  })

  it('shows front page or placeholder when switching to split screen', () => {
    let traits = {
      screen: { horizontal: 'compact' },
      window: { horizontal: 'compact' },
    }
    props.navigator = template.navigator({
      traitCollection: (callback) => { callback(traits) },
    })
    props.pages = [template.pageModel({ isFrontPage: false })]
    const course = props.course = template.courseModel({ id: '1' })
    const tree = shallow(<PagesList {...props} />)
    traits = {
      screen: { horizontal: 'regular' },
      window: { horizontal: 'regular' },
    }
    tree.find('Screen').simulate('TraitCollectionChange')
    expect(props.navigator.show).toHaveBeenLastCalledWith(
      '/courses/1/placeholder',
      {},
      { courseColor: 'green', course: course.raw },
    )
  })
})
