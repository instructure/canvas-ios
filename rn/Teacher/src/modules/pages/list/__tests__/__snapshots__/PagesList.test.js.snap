// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PagesList renders empty 1`] = `
<Screen
  navBarColor="green"
  navBarStyle="dark"
  onTraitCollectionChange={[Function]}
  rightBarButtons={
    Array [
      Object {
        "accessibilityLabel": "New Page",
        "action": [Function],
        "image": Object {
          "testUri": "../../../src/images/Add.png",
        },
        "testID": "pages.list.add.button",
      },
    ]
  }
  showDismissButton={true}
  statusBarStyle="default"
  subtitle="Learn React Native"
  title="Pages"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <FlatList
      ListEmptyComponent={
        <ListEmptyComponent
          title="There are no pages to display."
        />
      }
      data={Array []}
      extraData={
        Object {
          "pages": Array [],
          "selectedPageURL": null,
          "windowTraits": Object {
            "horizontal": "compact",
            "vertical": "regular",
          },
        }
      }
      keyExtractor={[Function]}
      onRefresh={[MockFunction]}
      refreshing={false}
      renderItem={[Function]}
      testID="pages.list.list"
    />
  </View>
</Screen>
`;

exports[`PagesList renders loading 1`] = `
<Screen
  navBarColor="green"
  navBarStyle="dark"
  onTraitCollectionChange={[Function]}
  rightBarButtons={
    Array [
      Object {
        "accessibilityLabel": "New Page",
        "action": [Function],
        "image": Object {
          "testUri": "../../../src/images/Add.png",
        },
        "testID": "pages.list.add.button",
      },
    ]
  }
  showDismissButton={true}
  statusBarStyle="default"
  title="Pages"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <FlatList
      ListEmptyComponent={null}
      data={Array []}
      extraData={
        Object {
          "pages": Array [],
          "selectedPageURL": null,
          "windowTraits": Object {
            "horizontal": "compact",
            "vertical": "regular",
          },
        }
      }
      keyExtractor={[Function]}
      onRefresh={[MockFunction]}
      refreshing={true}
      renderItem={[Function]}
      testID="pages.list.list"
    />
  </View>
</Screen>
`;

exports[`PagesList shows access icon for teachers 1`] = `
<View
  style={
    Object {
      "alignSelf": "flex-start",
    }
  }
>
  <AccessIcon
    disableAppSpecificChecks={false}
    entry={
      Object {
        "body": "<p>Hello, World!</p>",
        "created_at": "2017-03-17T19:15:25Z",
        "editing_roles": "teachers",
        "front_page": false,
        "hide_from_students": false,
        "html_url": "/pages/page-1",
        "page_id": "1",
        "published": true,
        "title": "Page 1",
        "updated_at": "2017-03-17T19:15:25Z",
        "url": "page-1",
      }
    }
    image={
      Object {
        "testUri": "../../../src/images/course/Pages.png",
      }
    }
    showAccessIcon={true}
    tintColor="green"
  />
</View>
`;
